### PR TITLE
Maid-1209_nfs separation

### PR DESCRIPTION
**Pull request by hitman401**
NFS code moved in from maidsafe_client. CI would pass only with maidsafe_client version 0.1.2 and above

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/maidsafe_nfs/4)

### CODE REVIEW

File: src/helper/directory_helper.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+                    },
+                    Err(_) => Err("IO Error".to_string())
+                }
+            },
+            Err(_) => Err("Encryption failed".to_string())
+        }
+    }
+
+    fn get_directory_version(&self, directory_id: &::routing::NameType,
+                            version: &::routing::NameType) -> Result<::directory_listing::DirectoryListing, String> {
```
Comment (ustulation) [inline src/helper/directory_helper.rs]: alignment

File: src/helper/file_helper.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+/// File provides helper functions to perform Operations on Files
+pub struct FileHelper {
+    client: ::std::sync::Arc<::std::sync::Mutex<::maidsafe_client::client::Client>>
```
Comment (ustulation) [inline src/helper/file_helper.rs]: put comma at the end for ease of addition (that is the style we will follow)

File: src/helper/file_helper.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+impl FileHelper {
+    /// Create a new FileHelper instance
+    pub fn new(client: ::std::sync::Arc<::std::sync::Mutex<::maidsafe_client::client::Client>>) -> FileHelper {
+        FileHelper {
+            client: client
```
Comment (ustulation) [inline src/helper/file_helper.rs]: comma at the end

File: src/helper/file_helper.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            Some(_) => {
+                file.get_mut_metadata().set_user_metadata(user_metadata.clone());
+                directory.upsert_file(file.clone());
+                let mut directory_helper = ::helper::DirectoryHelper::new(self.client.clone());
+                match directory_helper.update(&directory) {
+                    Ok(_) => Ok(()),
+                    Err(_) => Err("Failed to update".to_string())
+                }
+            },
+            None => Err("File not present in the directory".to_string())
```
Comment (ustulation) [inline src/helper/file_helper.rs]: comma at the end

File: src/metadata.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /// User setteble metadata for custom metadata
+    pub fn set_user_metadata(&mut self, user_metadata: Vec<u8>) {
+        self.user_metadata = user_metadata;
+    }
+
+    /// Get user setteble custom metadata
+    pub fn get_user_metadata(&self) -> Option<&Vec<u8>> {
+        if self.user_metadata.is_empty() {
```
Comment (ustulation) [inline src/metadata.rs:62]: `self.user_metadata.and(Some(&self.user_metadata))`

File: src/metadata.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    fn decode<D: Decoder>(d: &mut D)->Result<Metadata, D::Error> {
+        try!(d.read_u64());
+        let (name, size, meta, created_sec, created_nsec, modified_sec,
+            modified_nsec) = try!(Decodable::decode(d));
+
+        Ok(Metadata {
+                name: name,
+                user_metadata: meta,
+                size: size,
+                created_time:  ::time::at_utc(::time::Timespec {
```
Comment (ustulation) [inline src/metadata.rs]: extra space to be deleted

File: src/rest/blob.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    /// This can also be modified on the fly as the return is a mutable value
+    pub fn convert_to_mut_file(&mut self) -> &mut ::file::File {
+        &mut self.file
+    }
+
+    /// Convert the format acceptable to the lower level Api's into a Blob for more restful
+    /// interface
+    pub fn convert_from_file(file: ::file::File) -> Blob {
+        Blob {
+            file: file
```
Comment (ustulation) [inline src/rest/blob.rs]: comma at the end

File: src/rest/container.rs [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+        }
+    }
+
+    /// Returns the Created time of the container
+    pub fn get_created_time(&self) -> ::time::Tm {
+        self.directory_listing.get_metadata().get_created_time()
+    }
+
+    /// Return the unique id of the container
+    pub fn get_id(&self) -> [u8;64] {
```
Comment (ustulation) [inline src/rest/container.rs]: `[u8; 64]` gap

File: src/rest/container_info.rs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+/// Wrapper over DirectoryInfo to present Rest-friendly name to the Restful interface users
+pub struct ContainerInfo {
+    info: ::directory_info::DirectoryInfo,
+}
+
+impl ContainerInfo {
+    /// Get Container ID. This is the directory ID which is unique for every directory and is the
+    /// only way to retrieve that directory (DirectoryListing) from the network
+    pub fn get_id(&self) -> [u8;64] {
```
Comment (ustulation) [inline src/rest/container_info.rs]: gap after `;`

File: src/utils/mod.rs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+pub fn deserialise<T>(data: Vec<u8>) -> T where T : Decodable {
+    let mut d = cbor::Decoder::from_bytes(data);
+    d.decode().next().unwrap().unwrap()
+    // match d.decode().next().unwrap().unwrap() {
+    //     ::data_parser::Parser::StructuredData(obj) => obj,
+    //     ::data_parser::Parser::ImmutableData(obj) => obj,
+    // }
+}
+
+pub mod test {
```
Comment (ustulation) [inline src/utils/mod.rs]: why public ? And if requires to be public then should be in client

File: src/rest/container.rs [hunk 11]
```diff
@@ -0,0 +1,506 @@
...
+        }
+    }
+
+    fn list_container_versions(&mut self, container_id: &::routing::NameType) -> Result<Vec<[u8;64]>, String> {
+        let mut directory_helper = ::helper::DirectoryHelper::new(self.client.clone());
+        match directory_helper.get_versions(container_id) {
+            Ok(versions) => {
+                Ok(versions.iter().map(|v| v.0).collect())
+            },
+            Err(msg) => Err(msg)
```
Comment (ustulation) [inline src/rest/container.rs]: comma at the end

File: src/rest/container.rs [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+        self.directory_listing.get_metadata().get_name()
+    }
+
+    /// Returns the list of Blobs in the container
+    pub fn get_blobs(&self) -> Vec<::rest::Blob> {
+        self.directory_listing.get_files().iter().map(|x| ::rest::Blob::convert_from_file(x.clone())).collect()
+    }
+
+    /// Returns a Blob from the container
+    pub fn get_blob(&self, name: String, version: Option<[u8;64]>) -> Result<::rest::Blob, String> {
```
Comment (ustulation) [inline src/rest/container.rs]: gap after `;`'

File: src/metadata.rs (hunk 5 above)
Comment (hitman401) [inline src/metadata.rs:62]: As of now user_metadata is vec<u8>. We can use combinators once we make it as Option. We have it as a todo probably by end of sprint, if time permits

**Comment by ustulation** [comment]
Reviewed 4 of 19 files at r1, 13 of 14 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/maidsafe_nfs/4)

**Pull request merged**